File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length, dp[] = new int[n];

        for(int i = 0; i < n; ++i){
            dp[i] = 1;
            for(int j = i - 1; j >= 0; --j){
                if(nums[j] < nums[i] && dp[i] < 1 + dp[j]){
                    dp[i] = 1 + dp[j];
                }
            }
        }

        return Arrays.stream(dp).max().getAsInt();
    }
}

File: Data Structures & Algorithms/longest-common-subsequence/submission-2.py
class Solution:
    def longestCommonSubsequence(self, s: str, t: str) -> int:
        """
        if chars are equal
        fn(i,j) = 1 + fn(i-1, j-1)
        else
        fn(i,j) = max(fn(i-1,j), fn(i,j-1))

        in terms of arrays 
        equal case
        dp[j] = 1 + dp[j-1]
        else case
        dp[j] = max(dp[j], dp[j-1])
        """
        if(len(s) < len(t)):
            return self.longestCommonSubsequence(t,s)

        m, n = len(s), len(t)
        dp = [0] * (n+1)

        for i in range(m-1, -1, -1):
            prev = 0
            for j in range(n-1, -1, -1):
                temp = dp[j]
                if s[i] == t[j]:
                    dp[j] = 1 + prev
                else:
                    dp[j] = max(dp[j], dp[j+1])
                prev = temp
        return dp[0]

        